pkg/querier/frontend/frontend_querier_queues_test.go

```diff
@@ -0,0 +1,249 @@
+	assert.InDelta(t, stdDev, 0, mean*0.2)
+}
+
+func TestQueuesConsistency(t *testing.T) {
+	uq := newUserQueues(0)
+	assert.NotNil(t, uq)
+	assert.NoError(t, isConsistent(uq))
+
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+
+	lastUserIndexes := map[string]int{}
+
+	for i := 0; i < 1000; i++ {
+		switch r.Int() % 6 {
+		case 0:
+			assert.NotNil(t, uq.getOrAddQueue(generateTenant(r), 3))
+		case 1:
+			qid := generateQuerier(r)
+			_, _, luid := uq.getNextQueueForQuerier(lastUserIndexes[qid], qid)
+			lastUserIndexes[qid] = luid
+		case 2:
+			uq.deleteQueue(generateTenant(r))
+		case 3:
+			uq.addQuerierConnection(generateQuerier(r))
+		case 4:
+			uq.removeQuerierConnection(generateQuerier(r))
+		}
+
+		assert.NoErrorf(t, isConsistent(uq), "last action %d", i)
+	}
+}
+
+func generateTenant(r *rand.Rand) string {
+	return fmt.Sprint("tenant-", r.Int()%5)
+}
+
+func generateQuerier(r *rand.Rand) string {
+	return fmt.Sprint("querier-", r.Int()%5)
+}
+
+func getOrAdd(t *testing.T, uq *queues, tenant string, maxQueriers int) chan *request {
+	q := uq.getOrAddQueue(tenant, maxQueriers)
+	assert.NotNil(t, q)
+	assert.NoError(t, isConsistent(uq))
+	assert.Equal(t, q, uq.getOrAddQueue(tenant, maxQueriers))
+	return q
+}
+
+func confirmOrderForQuerier(t *testing.T, uq *queues, querier string, lastUserIndex int, qs ...chan *request) int {
+	var n chan *request
+	for _, q := range qs {
+		n, _, lastUserIndex = uq.getNextQueueForQuerier(lastUserIndex, querier)
+		assert.Equal(t, q, n)
+		assert.NoError(t, isConsistent(uq))
+	}
+	return lastUserIndex
+}
+
+func isConsistent(uq *queues) error {
+	if len(uq.sortedQueriers) != len(uq.querierConnections) {
+		return fmt.Errorf("inconsistent number of sorted queriers and querier connections")
+	}
+
+	uc := 0
+	for ix, u := range uq.users {
+		q := uq.userQueues[u]
+		if u != "" && q == nil {
+			return fmt.Errorf("user %s doesn't have queue", u)
+		}
+		if u == "" && q != nil {
+			return fmt.Errorf("user %s shouldn't have queue", u)
+		}
+		if u == "" {
+			continue
+		}
+
+		uc++
+
+		if q.index != ix {
+			return fmt.Errorf("invalid user's index, expected=%d, got=%d", ix, q.index)
+		}
+
+		if q.maxQueriers == 0 && q.queriers != nil {
+			return fmt.Errorf("user %s has queriers, but maxQueriers=0", u)
+		}
+
+		if q.maxQueriers > 0 && len(uq.sortedQueriers) <= q.maxQueriers && q.queriers != nil {
+			return fmt.Errorf("user %s has queriers set despite not enough queriers available", u)
+		}
+
+		if q.maxQueriers > 0 && len(uq.sortedQueriers) > q.maxQueriers && len(q.queriers) != q.maxQueriers {
+			return fmt.Errorf("user %s has incorrect number of queriers, expected=%d, got=%d", u, len(q.queriers), q.maxQueriers)
+		}
+	}
+
+	if uc != len(uq.userQueues) {
+		return fmt.Errorf("inconsistent number of users list and user queues")
+	}
+
+	return nil
+}
+
+func TestShuffleQueriers(t *testing.T) {
+	allQueriers := []string{"a", "b", "c", "d", "e"}
+
```

Comment:
Might be good to add a test that result of shuffleQueriersForUser has unique values and all were in the original input?